### ISSUE TITLE
MoltenVK doesn't handle stencil views of depth-stencil image

### ISSUE THREAD

**Issue by baldurk** [KhronosGroup/MoltenVK#243]
Vulkan lets you create a stencil view of a depth-stencil image to sample the stencil. You do so by specifying `VkImageViewCreateInfo::subresourceRange.aspectMask == VK_IMAGE_ASPECT_STENCIL_BIT` on the image view, then when sampled as a unsigned integer texture the view will return stencil in red instead of depth in red.

This seems like it can be mapped to metal, albeit with similar sort of swizzling restrictions to other formats. Something like the following proof-of-concept hack allows me to read from red in my stencil view successfully:

```patch
diff --git a/MoltenVK/MoltenVK/GPUObjects/MVKImage.h b/MoltenVK/MoltenVK/GPUObjects/MVKImage.h
index 48e0e53..8b1cbef 100644
--- a/MoltenVK/MoltenVK/GPUObjects/MVKImage.h
+++ b/MoltenVK/MoltenVK/GPUObjects/MVKImage.h
@@ -277,7 +277,7 @@ public:
 protected:
 	id<MTLTexture> newMTLTexture();
 	void initMTLTextureViewSupport();
-    MTLPixelFormat getSwizzledMTLPixelFormat(VkFormat format, VkComponentMapping components);
+    MTLPixelFormat getSwizzledMTLPixelFormat(VkFormat format, VkImageAspectFlags aspect, VkComponentMapping components);
     bool matchesSwizzle(VkComponentMapping components, VkComponentMapping pattern);
     const char* getSwizzleName(VkComponentSwizzle swizzle);
     void setSwizzleFormatError(VkFormat format, VkComponentMapping components);
diff --git a/MoltenVK/MoltenVK/GPUObjects/MVKImage.mm b/MoltenVK/MoltenVK/GPUObjects/MVKImage.mm
index 6b84a1b..64b446b 100644
--- a/MoltenVK/MoltenVK/GPUObjects/MVKImage.mm
+++ b/MoltenVK/MoltenVK/GPUObjects/MVKImage.mm
@@ -464,7 +464,7 @@ MVKImage::MVKImage(MVKDevice* device, const VkImageCreateInfo* pCreateInfo) : MV
 
     _isDepthStencilAttachment = (mvkAreFlagsEnabled(pCreateInfo->usage, VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT) ||
                                  mvkAreFlagsEnabled(mvkVkFormatProperties(pCreateInfo->format).optimalTilingFeatures, VK_FORMAT_FEATURE_DEPTH_STENCIL_ATTACHMENT_BIT));
-	_canSupportMTLTextureView = !_isDepthStencilAttachment;
+	_canSupportMTLTextureView = true;//!_isDepthStencilAttachment;
     _hasExpectedTexelSize = (mvkMTLPixelFormatBytesPerBlock(_mtlPixelFormat) == mvkVkFormatBytesPerBlock(pCreateInfo->format));
 	_isLinear = validateLinear(pCreateInfo);
 	_usesTexelBuffer = false;
@@ -628,7 +628,7 @@ MVKImageView::MVKImageView(MVKDevice* device, const VkImageViewCreateInfo* pCrea
 	}
 
 	_mtlTexture = nil;
-    _mtlPixelFormat = getSwizzledMTLPixelFormat(pCreateInfo->format, pCreateInfo->components);
+    _mtlPixelFormat = getSwizzledMTLPixelFormat(pCreateInfo->format, _subresourceRange.aspectMask, pCreateInfo->components);
 	_mtlTextureType = mvkMTLTextureTypeFromVkImageViewType(pCreateInfo->viewType, (_image->getSampleCount() != VK_SAMPLE_COUNT_1_BIT));
 	initMTLTextureViewSupport();
 }
@@ -649,10 +649,18 @@ void MVKImageView::validateImageViewConfig(const VkImageViewCreateInfo* pCreateI
 // Metal does not support general per-texture swizzles, and so this function relies on a few coincidental
 // alignments of existing MTLPixelFormats of the same structure. If swizzling is not possible for a
 // particular combination of format and swizzle spec, the original MTLPixelFormat is returned.
-MTLPixelFormat MVKImageView::getSwizzledMTLPixelFormat(VkFormat format, VkComponentMapping components) {
+MTLPixelFormat MVKImageView::getSwizzledMTLPixelFormat(VkFormat format, VkImageAspectFlags aspect, VkComponentMapping components) {
     MTLPixelFormat mtlPF = mtlPixelFormatFromVkFormat(format);
 
     switch (mtlPF) {
+        case MTLPixelFormatDepth32Float_Stencil8:
+            if (aspect == VK_IMAGE_ASPECT_STENCIL_BIT)
+                return MTLPixelFormatX32_Stencil8;
+            break;
+        case MTLPixelFormatDepth24UNorm_Stencil8:
+            if (aspect == VK_IMAGE_ASPECT_STENCIL_BIT)
+                return MTLPixelFormatX24_Stencil8;
+            break;
         case MTLPixelFormatR8Unorm:
             if (matchesSwizzle(components, {VK_COMPONENT_SWIZZLE_ZERO, VK_COMPONENT_SWIZZLE_MAX_ENUM, VK_COMPONENT_SWIZZLE_MAX_ENUM, VK_COMPONENT_SWIZZLE_R} ) ) {
                 return MTLPixelFormatA8Unorm;
```

I also noticed the `_canSupportMTLTextureView` flag which is set to false for depth-stencil attachments. I guess this restriction doesn't apply to casting between compatible depth-stencil formats because forcing that to true as above worked for me, so this would need a more fine-grained feature check.

**Comment by oscarbg** [comment]
+1

**Comment by billhollings** [comment]
@baldurk Are you able to package and test your changes...and submit them as a PR here?

**Comment by baldurk** [comment]
I didn't really want to make a PR since I don't know anything about metal and I don't know the implications of the `_canSupportMTLTextureView` change in particular. I'm not doing any more testing than making sure renderdoc can display stencil in a texture.

**Comment by cdavis5e** [comment]
It's funny you should mention this, because I just hit the point in `vkd3d`'s D3D12 tests where we now need to get a stencil-only view on a depth/stencil texture. Unfortunately, we need stencil in green, not red. I think I can fix that, but it's going to be a bit more involved.

Anyway, I cleaned up your changes and added a check that the stencil view formats are supported (requires Metal 1.2). See #252.

**Comment by baldurk** [comment]
Thanks for setting up the PR! Yes fortunately I'm able to side-step the need for green because I can change my shader code, it will be more annoying for D3D emulation and might need shader patching I suppose.

**Comment by billhollings** [comment]
@baldurk :

@cdavis5e addressed this in PR #252. Can you retest and close this issue if suitable.

**Comment by baldurk** [comment]
Works as expected, thanks!
